packages/scheduler/src/forks/SchedulerHostConfig.default.js

```diff
@@ -244,103 +207,162 @@ if (
-        nextFrameTime < previousFrameTime ? previousFrameTime : nextFrameTime;
-    } else {
-      previousFrameTime = nextFrameTime;
-    }
-    frameDeadline = rafTime + activeFrameTime;
-    if (!isMessageEventScheduled) {
-      isMessageEventScheduled = true;
-      port.postMessage(undefined);
+      prevRAFInterval = rAFInterval;
     }
+    prevRAFTime = rAFTime;
+    frameDeadline = rAFTime + frameLength;
+
+    port.postMessage(null);
   };
 
   requestHostCallback = function(callback) {
-    if (scheduledHostCallback === null) {
-      scheduledHostCallback = callback;
-      if (!isAnimationFrameScheduled) {
-        // If rAF didn't already schedule one, we need to schedule a frame.
-        // TODO: If this rAF doesn't materialize because the browser throttles,
-        // we might want to still have setTimeout trigger rIC as a backup to
-        // ensure that we keep performing work.
-        isAnimationFrameScheduled = true;
-        requestAnimationFrameWithTimeout(animationTick);
+    scheduledHostCallback = callback;
+    if (!isRAFLoopRunning) {
+      isRAFLoopRunning = true;
+
+      // Start a rAF loop.
+      requestAnimationFrame(rAFTime => {
+        if (requestIdleCallbackBeforeFirstFrame) {
+          cancelIdleCallback(idleCallbackID);
+        }
+        if (requestTimerEventBeforeFirstFrame) {
+          clearTimeout(idleTimeoutID);
+        }
+        onAnimationFrame(rAFTime);
+      });
+
+      // If we just missed the last vsync, the next rAF might not happen for
+      // another frame. To claim as much idle time as possible, post a callback
+      // with `requestIdleCallback`, which should fire if there's idle time left
+      // in the frame.
+      //
+      // This should only be an issue for the first rAF in the loop; subsequent
+      // rAFs are scheduled at the beginning of the preceding frame.
+      let idleCallbackID;
+      if (requestIdleCallbackBeforeFirstFrame) {
+        idleCallbackID = requestIdleCallback(() => {
+          if (requestTimerEventBeforeFirstFrame) {
+            clearTimeout(idleTimeoutID);
+          }
+          frameDeadline = getCurrentTime() + frameLength;
+          performWorkUntilDeadline();
+        });
+      }
+      // Alternate strategy to address the same problem. Scheduler a timer with
+      // no delay. If this fires before the rAF, that likely indicates that
+      // there's idle time before the next vsync. This isn't always the case,
+      // but we'll be aggressive and assume it is, as a trade off to prevent
+      // idle periods.
+      let idleTimeoutID;
+      if (requestTimerEventBeforeFirstFrame) {
+        idleTimeoutID = setTimeout(() => {
+          if (requestIdleCallbackBeforeFirstFrame) {
+            cancelIdleCallback(idleCallbackID);
+          }
+          frameDeadline = getCurrentTime() + frameLength;
+          performWorkUntilDeadline();
+        }, 0);
       }
     }
   };
 
   cancelHostCallback = function() {
```

Comment:
Worth checking/clearing the timeout/rIC here?

Reply:
`performWorkUntilDeadline` will exit immediately when `scheduledHostCallback` is null so it's fine. I wanted to keep the callback/timeout IDs local.

Reply:
Yeah, I know it's a no-op in that case.

Seems reasonable.